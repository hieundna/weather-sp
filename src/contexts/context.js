import { createContext, useCallback, useContext, useReducer } from 'react';
import { rootReducer, initialState } from './reducers/rootReducer'

const Context = createContext({});

export const AppProvider = props => {
    const { children } = props;
    const [data, dispatch] = useReducer(rootReducer, initialState);
    return (
        <Context.Provider value={{ data, setData: dispatch }}>
            {children}
        </Context.Provider>
    )
}

export const useAppContext = () => {
    const { data, setData } = useContext(Context)

    const dispatch = useCallback(
        ({ type, payload }) => {
            setData({ payload, type })
        },
        [setData]
    )
    return [data, dispatch]
}

export default Context;