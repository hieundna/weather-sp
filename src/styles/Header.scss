.header {
    display: flex;
    justify-content: space-between;
    max-width: 100%;
    height: 40px;
    padding: 20px 25px;

    .burger {
        width: 30px;
        display: inline-block;
    }

    .burger span {
        display: block;
        width: 100%;
        height: 3px;
        background-color: #fff;
        margin: 6px 0;
        transition: .4s;
    }

    .title {
        text-align: center;
        position: relative;
        cursor: pointer;
        width: 200px;
        .city-name {
            margin-right: 10px;
        }
    }
    .dropdown {
        display: none;
        width: 200px;
        left: 0px;
        background-color: rgb(73, 73, 73);
        position: absolute;
        bottom: -90px;
        .option {
            padding: 3px 10px;
        }
        .option:hover {
            background-color: rgba(255,255,255,0.3);
        }
    }
    .show {
        display: block;
    }
}

@media screen and (max-height: 694px){
    .header {
        height: 32px;
        padding: 8px 10px;
    }
}