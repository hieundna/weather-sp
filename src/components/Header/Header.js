import '../../styles/Header.scss';
// import Context from '../../contexts/context'
import { getWeatherData } from '../../api/index';
import { getWeatherInfor } from '../../contexts/actions/action';
import { useEffect, useState } from 'react';
import { useAppContext } from '../../contexts/context';

function Header() {
    const [stateData, dispatch] = useAppContext();
    const [selectedCity, setSelectedCity] = useState('Singapore');

    const [show, setShow] = useState(false);

    useEffect(async () => {
        window.addEventListener('click', windowClick);
        return () => {
            window.removeEventListener('click', windowClick);
        }
    }, [])

    const selectCity = async (index) => {
        setShow(!show);
        try {
            const data = await getWeatherData(stateData.city[index]);
            dispatch(getWeatherInfor(data.data));
            setSelectedCity(stateData.city[index]);
        } catch (error) {
            console.error('Error city:', error);
        }
    }

    const windowClick = e => {
        if (!e.target.classList.contains('dropdown')
            && !e.target.classList.contains('option')
            && !e.target.classList.contains('list-city')
            && !e.target.classList.contains('city-name')
            && !e.target.classList.contains('fal')) {
            setShow(false);
            return;
        }
    };

    return (
        <div className='header'>
            <div className='burger'>
                <span></span>
                <span></span>
                <span></span>
            </div>
            <div className='title'>
                <div role='myENV-test'>myENV</div>
                <div className='list-city' onClick={() => setShow(!show)}>
                    <span className='city-name'>{selectedCity}</span>
                    <i className="fal fa-angle-down fa-lg"></i>
                </div>
                <div className={`dropdown ${show ? 'show' : ''}`}>
                    {stateData.city?.map((cityName, idx) =>
                        <div key={idx}
                            className='option'
                            onClick={() => selectCity(idx)}
                        >
                            {cityName}
                        </div>
                    )}
                </div>
            </div>
            <div><i className="fas fa-bell fa-2x"></i></div>
        </div>
    );
}

export default Header;
