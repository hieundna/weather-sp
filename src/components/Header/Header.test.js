import React from 'react';
import Header from './Header.js';
import { render } from '@testing-library/react';
import { AppProvider } from '../../contexts/context'

const setup = () => render(<AppProvider> <Header /> </AppProvider>);
describe('Header Component test', () => {
    test("expect to render Header component", () => {
        const wrapper = setup();
        const title = wrapper.getByRole('myENV-test');
        expect(title).toHaveTextContent('myENV');
        expect(wrapper).toMatchSnapshot();
    });
})