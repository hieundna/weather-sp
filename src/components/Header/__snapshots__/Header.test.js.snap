// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component test expect to render Header component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
       
      <div
        class="header"
      >
        <div
          class="burger"
        >
          <span />
          <span />
          <span />
        </div>
        <div
          class="title"
        >
          <div
            role="myENV-test"
          >
            myENV
          </div>
          <div
            class="list-city"
          >
            <span
              class="city-name"
            >
              Singapore
            </span>
            <i
              class="fal fa-angle-down fa-lg"
            />
          </div>
          <div
            class="dropdown "
          >
            <div
              class="option"
            >
              Ho Chi Minh
            </div>
            <div
              class="option"
            >
              Vinh
            </div>
            <div
              class="option"
            >
              Singapore
            </div>
          </div>
        </div>
        <div>
          <i
            class="fas fa-bell fa-2x"
          />
        </div>
      </div>
       
    </div>
  </body>,
  "container": <div>
     
    <div
      class="header"
    >
      <div
        class="burger"
      >
        <span />
        <span />
        <span />
      </div>
      <div
        class="title"
      >
        <div
          role="myENV-test"
        >
          myENV
        </div>
        <div
          class="list-city"
        >
          <span
            class="city-name"
          >
            Singapore
          </span>
          <i
            class="fal fa-angle-down fa-lg"
          />
        </div>
        <div
          class="dropdown "
        >
          <div
            class="option"
          >
            Ho Chi Minh
          </div>
          <div
            class="option"
          >
            Vinh
          </div>
          <div
            class="option"
          >
            Singapore
          </div>
        </div>
      </div>
      <div>
        <i
          class="fas fa-bell fa-2x"
        />
      </div>
    </div>
     
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
